File: mvrappor.py
import numpy as np
import pandas as pd
import hashlib
from sklearn.linear_model import Lasso
from scipy.optimize import lsq_linear

# --- Parameters ---
num_bloombits = 2048
num_hashes = 5
prob_p = 0.5
prob_q = 0.75
prob_f = 0.5
num_views = 15
max_url_length = 100
lambda_lasso = 0.0001

# --- Load and trim dataset (first 1000 rows) ---
data = pd.read_csv('PhiUSIIL_Phishing_URL_Dataset.csv', header=None)
urls = data.iloc[1:, 1].astype(str).values
labels = data.iloc[1:, -1].astype(int).values
N = min(1000, len(urls))
websites = urls[:N]
labels = labels[:N]

# --- Helper functions ---
def bloom_hash(input_string, salt, num_bits):
    key = (input_string + "_" + salt).encode('utf-8')
    hash_digest = hashlib.sha256(key).digest()
    hash_val = int.from_bytes(hash_digest[:4], 'big')
    return hash_val % num_bits

def pad_string(s, maxlen):
    s = s[:maxlen].ljust(maxlen)
    return s

def normalize_urls(urls):
    urls = [u.lower().replace("https://", "").replace("http://", "").rstrip("/") for u in urls]
    return urls

# --- Step 1: PRR Encoding ---
prr_matrix = np.zeros((N, num_bloombits), dtype=int)
for i in range(N):
    s = pad_string(websites[i], max_url_length)
    bf = np.zeros(num_bloombits, dtype=int)
    for h in range(1, num_hashes + 1):
        salt = f"{h}_cohort"
        idx = bloom_hash(s, salt, num_bloombits)
        bf[idx] = 1
    f_mask = np.random.rand(num_bloombits) < prob_f
    uniform = np.random.rand(num_bloombits) < 0.5
    prr = np.where(f_mask, uniform, bf)
    prr_matrix[i] = prr

# --- Step 2: IRR Views ---
irr_matrix = []
for v in range(num_views):
    print(f"View {v+1}")
    for i in range(N):
        prr = prr_matrix[i]
        p_bits = np.random.rand(num_bloombits) < prob_p
        q_bits = np.random.rand(num_bloombits) < prob_q
        irr = np.where(prr == 1, q_bits, p_bits)
        irr_matrix.append(irr)
irr_matrix = np.array(irr_matrix)

# --- Step 3: Design Matrix for Candidates ---
candidates = np.unique(websites)
M = len(candidates)
A = np.zeros((M, num_bloombits), dtype=int)
for i, s in enumerate(candidates):
    s = pad_string(s, max_url_length)
    for h in range(1, num_hashes + 1):
        salt = f"{h}_cohort"
        idx = bloom_hash(s, salt, num_bloombits)
        A[i, idx] = 1

from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

# Compute y as usual
y = irr_matrix.mean(axis=0)

# Compute alpha correctly
n_samples = A.shape[0]
alpha = lambda_lasso / n_samples

# Setup Lasso pipeline with standardization
model = make_pipeline(
    StandardScaler(),
    Lasso(
        alpha=alpha,
        fit_intercept=True,
        positive=False,
        tol=1e-4,
        max_iter=2000
    )
)

try:
    model.fit(A.T, y)
    # Extract Lasso from pipeline to get coefficients
    lasso_coef = model.named_steps['lasso'].coef_
    reconstructed_freq = np.maximum(lasso_coef, 0)
    nonzero = np.sum(reconstructed_freq > 0)
    print(f"Used LASSO: {nonzero} non-zero coefficients")
    if nonzero == 0:
        raise Exception("Fallback to lsqnonneg")
except:
    res = lsq_linear(A.T, y, bounds=(0, np.inf))
    reconstructed_freq = res.x
    print(f"Used lsqnonneg: {np.sum(reconstructed_freq > 0)} non-zero coefficients")


# --- Step 5: Detection with Normalized Matching ---
threshold = 0.003
detected = np.where(reconstructed_freq > threshold)[0]
recon_websites = candidates[detected]

recon_clean = normalize_urls(recon_websites)
true_malicious = normalize_urls(websites[labels == 0])
recon_labels = [1 if r in true_malicious else 0 for r in recon_clean]

print("\nReconstructed Websites (label = 1 = phishing):")
for i, r in enumerate(recon_websites):
    print(f"{r}\tlabel={recon_labels[i]}")

# --- Metrics ---
tp = sum(recon_labels)
fp = len(recon_labels) - tp
fn = sum([1 for url in true_malicious if url not in recon_clean])
precision = tp / (tp + fp + 1e-10)
recall = tp / (tp + fn + 1e-10)
f1 = 2 * precision * recall / (precision + recall + 1e-10)
print(f"\n# Detected: {len(recon_websites)}\n# True phishing: {len(true_malicious)}\n# Matched: {tp}")
print(f"Precision: {precision:.2f} | Recall: {recall:.2f} | F1-score: {f1:.2f}")

# --- Top Results ---
top_idx = np.argsort(reconstructed_freq)[::-1]
print("\nTop Candidates by Estimated Frequency:")
for i in range(min(10, len(top_idx))):
    print(f"{candidates[top_idx[i]]}: {reconstructed_freq[top_idx[i]]:.4f}")

File: mvrappor.m
clear all; clc;

% --- Parameters ---
num_bloombits = 6144;
num_hashes = 1;
prob_p = 0.3;
prob_q = 0.95;
prob_f = 0.4;
num_views = 55;
max_url_length = 100;
lambda_lasso = 1e-5;

% --- Load and trim dataset (first 1000 rows) ---
data = readtable('PhiUSIIL_Phishing_URL_Dataset.csv', 'ReadVariableNames', false);
urls = string(data{2:end, 2});
labels = double(data{2:end, end});
N = min(1000, length(urls));
websites = urls(1:N);
labels = labels(1:N);

% --- Step 1: PRR Encoding (preallocated) ---
prr_matrix = false(N, num_bloombits);
for i = 1:N
    s = pad_string(websites(i), max_url_length);
    bf = false(1, num_bloombits);
    for h = 1:num_hashes
        salt = strcat(num2str(h), '_cohort');
        idx = bloom_hash(s, salt, num_bloombits);
        bf(idx) = 1;
    end
    f_mask = rand(1, num_bloombits) < prob_f;
    uniform = rand(1, num_bloombits) < 0.5;
    prr = (bf & ~f_mask) | (uniform & f_mask);
    prr_matrix(i, :) = prr;
end

% --- Step 2: IRR Views (preallocated, fast) ---
irr_matrix = false(N * num_views, num_bloombits);
for v = 1:num_views
    % Vectorized bit generation
    p_bits = rand(N, num_bloombits) < prob_p;
    q_bits = rand(N, num_bloombits) < prob_q;
    prr_view = prr_matrix;
    irr = (p_bits & ~prr_view) | (q_bits & prr_view);
    irr_matrix((v-1)*N + 1:v*N, :) = irr;
end

% --- Step 3: Design Matrix for Candidates (preallocated) ---
candidates = unique(websites);
M = length(candidates);
A = false(M, num_bloombits);
for i = 1:M
    s = pad_string(candidates(i), max_url_length);
    for h = 1:num_hashes
        salt = strcat(num2str(h), '_cohort');
        idx = bloom_hash(s, salt, num_bloombits);
        A(i, idx) = 1;
    end
end

% --- Step 4: Frequency Reconstruction ---
y = mean(irr_matrix, 1)';
try
    [B, ~] = lasso(double(A'), y, 'Lambda', lambda_lasso);
    reconstructed_freq = max(B(:, 1), 0);
    fprintf('Used LASSO: %d non-zero coefficients\n', sum(reconstructed_freq > 0));
    if sum(reconstructed_freq > 0) == 0
        error('Fallback to lsqnonneg');
    end
catch
    reconstructed_freq = lsqnonneg(double(A'), y);
    fprintf('Used lsqnonneg: %d non-zero coefficients\n', sum(reconstructed_freq > 0));
end

% --- Step 5: Detection with Normalized Matching ---
threshold = 5e-07;
detected = find(reconstructed_freq > threshold);
recon_websites = candidates(detected);

% Normalize both sets for string comparison
recon_clean = normalize_urls(recon_websites);
true_malicious = normalize_urls(websites(labels == 0));
recon_labels = ismember(recon_clean, true_malicious);

fprintf('\nReconstructed Websites (label = 1 = phishing):\n');
for i = 1:length(recon_websites)
    fprintf('%s\tlabel=%d\n', recon_websites(i), recon_labels(i));
end

% --- Metrics ---
tp = sum(recon_labels);
fp = length(recon_labels) - tp;
fn = sum(~ismember(true_malicious, recon_clean));
precision = tp / (tp + fp + eps);
recall = tp / (tp + fn + eps);
f1 = 2 * precision * recall / (precision + recall + eps);
fprintf('\n# Detected: %d\n# True phishing: %d\n# Matched: %d\n', ...
        length(recon_websites), length(true_malicious), tp);
fprintf('Precision: %.2f | Recall: %.2f | F1-score: %.2f\n', precision, recall, f1);

% --- Top Results ---
[~, idx] = sort(reconstructed_freq, 'descend');
fprintf('\nTop Candidates by Estimated Frequency:\n');
for i = 1:min(10, length(idx))
    fprintf('%s: %.4f\n', candidates(idx(i)), reconstructed_freq(idx(i)));
end

% --- Helpers ---
function index = bloom_hash(input_string, salt, num_bits)
    key = [char(input_string), '_', salt];
    md = java.security.MessageDigest.getInstance('SHA-256');
    hash = md.digest(uint8(key));
    hash_num = typecast(hash(1:4), 'uint32');
    index = mod(double(hash_num), num_bits) + 1;
end

function padded = pad_string(s, maxlen)
    s = char(s);
    padded = s(1:min(end, maxlen));
    padded = [padded, repmat(' ', 1, maxlen - length(padded))];
end

function out = normalize_urls(urls)
    urls = lower(urls);
    urls = regexprep(urls, '^https?://', '');
    urls = regexprep(urls, '/$', '');
    out = urls;
end
